### ISSUE TITLE
LightGBM breaks numpy's float matrix multiplication.

### ISSUE THREAD

**Issue by Akiiino** [Microsoft/LightGBM#1444]
I've got the weirdest bug. LightGBM seems to break `numpy`'s basic floating point math capabilities. The easiest way to reproduce this for me is this:

```
import numpy as np
import lightgbm

a = np.arange(10000).reshape(100,100).astype(float)*2
b = np.ones(100).astype(float)*2

x = a.dot(b) 
y = a.dot(b) 

print(np.all(x == y))
```

This promptly prints `False` 90% of the time. However, if I use `numpy`'s float matrix multiplication at least once before importing `lightgbm`, everything works just fine:

```
import numpy as np
np.array([1,2,3]).astype(float).dot(np.array([1,2,3]).astype(float))

import lightgbm

a = np.arange(10000).reshape(100,100).astype(float)*2
b = np.ones(100).astype(float)*2

x = a.dot(b) 
y = a.dot(b) 

print(np.all(x == y))
```

This always returns `True`. Everything also works fine if the matrices are not `float`.

I've been trying to locate the bug for two days now. I've gone from blaming `numpy` to blaming my python installation to blaming my whole system. Today I went and reinstalled the OS from scratch, without transferring anything over from the backup. The behaviour still persists. I'm at a loss as to what could possibly cause this beside some deeply arcane bug inside LightGBM.

## Environment info
Operating System: macOS 10.13.5
CPU: Intel Core i7 
C++/Python/R version: Python 3.6.5 from Anaconda, G++-8.1.0

**Comment by StrikerRUS** [comment]
@Akiiino Huh, really rather weird!

On my machine (Windows) both code pieces print `True` every time.

**Comment by Akiiino** [comment]
Yeah, I should have added that I've asked a couple of friends to try this snippet out for me, and they've reported no signs of any problem.

**Comment by StrikerRUS** [comment]
OK. I've copied your first snippet as one of our tests:
https://github.com/Microsoft/LightGBM/blob/0b454c46357f90036b8be6fc68b1ad60baf2ebcd/tests/python_package_test/test_sklearn.py#L39-L46

All 4 macOS jobs have no errors...
https://travis-ci.org/Microsoft/LightGBM/builds/390877799

**Comment by StrikerRUS** [comment]
And what about versions of packages?

At continuous integration service two versions of `numpy` have been tested: `1.11.3` and `1.14.3`.

**Comment by Akiiino** [comment]
```akiiinos-MacBook-Pro:~ akiiino$ python
Python 3.6.5 |Anaconda, Inc.| (default, Apr 26 2018, 08:42:37)
[GCC 4.2.1 Compatible Clang 4.0.1 (tags/RELEASE_401/final)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np
>>> np.__version__
'1.14.3'
>>> import lightgbm
>>> lightgbm.__version__
'2.1.1'
>>>
>>> a = np.arange(10000).reshape(100,100).astype(float)*2
>>> b = np.ones(100).astype(float)*2
>>>
>>> x = a.dot(b)
>>> y = a.dot(b)
>>>
>>> print(np.all(x == y))
False
>>>```

**Comment by StrikerRUS** [comment]
To be honest, I don't believe that this is the reason.

Does this issue occur only with import of LightGBM? Maybe you are able to check with importing XGBoost?

**Comment by Akiiino** [comment]
```Python 3.6.5 |Anaconda, Inc.| (default, Apr 26 2018, 08:42:37)
[GCC 4.2.1 Compatible Clang 4.0.1 (tags/RELEASE_401/final)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np
>>> import xgboost
>>>
>>> a = np.arange(10000).reshape(100,100).astype(float)*2
>>> b = np.ones(100).astype(float)*2
>>>
>>> x = a.dot(b)
>>> y = a.dot(b)
>>>
>>> print(np.all(x == y))
True
```

Yeah, I've only seen this happen in this specific configuration with LightGBM --- if this was easier to trigger, I would have noticed errors and bugs a long time ago, I think.

**Comment by StrikerRUS** [comment]
What happens if you use `numpy.allclose()` to compare arrays?

**Comment by StrikerRUS** [comment]
Also I recommend to duplicate this issue in numpy repo. Maybe they'll shed some light on this.

**Comment by Akiiino** [comment]
```
Python 3.6.5 |Anaconda, Inc.| (default, Apr 26 2018, 08:42:37)
[GCC 4.2.1 Compatible Clang 4.0.1 (tags/RELEASE_401/final)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np
>>> import lightgbm
>>>
>>> a = np.arange(10000).reshape(100,100).astype(float)*2
>>> b = np.ones(100).astype(float)*2
>>>
>>> x = a.dot(b)
>>> y = a.dot(b)
>>>
>>> x
array([   79200.,   239200.,   399200.,   559200.,   719200.,   879200.,
        1039200.,  1199200.,  1359200.,  1519200.,  1679200.,  1839200.,
        1999200.,  2159200.,  2319200.,  2479200.,  2639200.,  2799200.,
        2959200.,  3119200.,  3279200.,  3439200.,  3599200.,  3759200.,
        3919200.,  4079200.,  4239200.,  4399200.,  4559200.,  4719200.,
        4879200.,  5039200.,  5199200.,  5359200.,  5519200.,  5679200.,
        5839200.,  5999200.,  6159200.,  6319200.,  6479200.,  6639200.,
        6799200.,  6959200.,  7119200.,  7279200.,  7439200.,  7599200.,
        7759200.,  7919200.,  8079200.,  8239200.,  8399200.,  8559200.,
        8719200.,  8879200.,  9039200.,  9199200.,  9359200.,  9519200.,
        9679200.,  9839200.,  9999200., 10159200., 10319200., 10479200.,
       10639200., 10799200., 10959200., 11119200., 11279200., 11439200.,
       11599200., 11759200., 11919200., 12079200., 12239200., 12399200.,
       12559200., 12719200., 12879200., 13039200., 13199200., 13359200.,
       13519200., 13679200., 13839200., 13999200., 14159200., 14319200.,
       14479200., 14639200., 14799200., 14959200., 15119200., 15279200.,
       15439200., 15599200., 15759200., 15919200.])
>>> y
array([  19800.,   59800.,   99800.,  139800.,  179800.,  219800.,
        259800.,  299800.,  339800.,  379800.,  419800.,  459800.,
        499800.,  539800.,  579800.,  619800.,  659800.,  699800.,
        739800.,  779800.,  819800.,  859800.,  899800.,  939800.,
        979800., 1019800., 1059800., 1099800., 1139800., 1179800.,
       1219800., 1259800., 1299800., 1339800., 1379800., 1419800.,
       1459800., 1499800., 1539800., 1579800., 1619800., 1659800.,
       1699800., 1739800., 1779800., 1819800., 1859800., 1899800.,
       1939800., 1979800., 2019800., 2059800., 2099800., 2139800.,
       2179800., 2219800., 2259800., 2299800., 2339800., 2379800.,
       2419800., 2459800., 2499800., 2539800., 2579800., 2619800.,
       2659800., 2699800., 2739800., 2779800., 2819800., 2859800.,
       2899800., 2939800., 2979800., 3019800., 3059800., 3099800.,
       3139800., 3179800., 3219800., 3259800., 3299800., 3339800.,
       3379800., 3419800., 3459800., 3499800., 3539800., 3579800.,
       3619800., 3659800., 3699800., 3739800., 3779800., 3819800.,
       3859800., 3899800., 3939800., 3979800.])
```

Absolutely bizarre. Does not look like a float precision problem to me.

Yeah, I guess it can't hurt to post this there as well.

**Comment by StrikerRUS** [comment]
The similar issue have been posted yesterday in XGBoost repo. There, the line `import xgboost` causes segmentation fault in `scipy`'s `SVD`.
https://github.com/dmlc/xgboost/issues/3378

**Comment by StrikerRUS** [comment]
@Akiiino Can you please run your tests with

1. LightGBM + numpy from PyPI
2. LightGBM + numpy from conda

?

BTW, how do you install LightGBM? From sources or PyPI?

**Comment by StrikerRUS** [comment]
ping @Akiiino

**Comment by StrikerRUS** [comment]
Both LightGBM and XGBoost use OpenMP, so this may be the same issue: https://github.com/ContinuumIO/anaconda-issues/issues/8803
